### PR TITLE
[Issue 5520][Function Doc] Add window function context docs for repo

### PR DESCRIPTION
**Pull request by wolfstudy**
Signed-off-by: xiaolong.ran <rxl@apache.org>

Fixes: #5520

### Modifications

- Add window function context docs for repo

### CODE REVIEW

File: site2/docs/window-functions-context.md [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+Spec contains the basic information of a function.
+
+### Get input topics
+
+`getInputTopics` method gets the **name list** of all input topics.
+
+This example demonstrates how to get the name list of all input topics in a Java window function.
+
+<!--DOCUSAURUS_CODE_TABS-->
+<!--Java-->
```
Comment (sijia-w) [inline site2/docs/window-functions-context.md:49]: Delete lines 48-49. When support and examples in other languages are available, we'll have them back.

File: site2/docs/window-functions-context.md [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+<!--END_DOCUSAURUS_CODE_TABS-->
+
+### Get output topic
+
+`getOutputTopic` method gets the **name of a topic** to which the message is sent.
+
+This example demonstrates how to get the name of an output topic in a Java window function.
+
+<!--DOCUSAURUS_CODE_TABS-->
+<!--Java-->
```
Comment (sijia-w) [inline site2/docs/window-functions-context.md:72]: Same here. Lines 71-71 & 85.

File: site2/docs/window-functions-context.md [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+```
+<!--END_DOCUSAURUS_CODE_TABS-->
+
+### Get tenant
+
+`getTenant` method gets the tenant name associated with the window function.
+
+This example demonstrates how to get the tenant name in a Java window function.
+
+<!--DOCUSAURUS_CODE_TABS-->
```
Comment (sijia-w) [inline site2/docs/window-functions-context.md:93]: Same here.

File: site2/docs/window-functions-context.md [hunk 4]
```diff
@@ -0,0 +1,593 @@
...
+```
+<!--END_DOCUSAURUS_CODE_TABS-->
+
+### Get namespace
+
+`getNamespace` method gets the namespace associated with the window function.
+
+This example demonstrates how to get the namespace in a Java window function.
+
+<!--DOCUSAURUS_CODE_TABS-->
```
Comment (sijia-w) [inline site2/docs/window-functions-context.md:116]: Same here.

File: site2/docs/window-functions-context.md [hunk 5]
```diff
@@ -0,0 +1,593 @@
...
+        return null;
+    }
+
+}
+```
+<!--END_DOCUSAURUS_CODE_TABS-->
+
+### Get output topic
+
+`getOutputTopic` method gets the **name of a topic** to which the message is sent.
```
Comment (sijia-w) [inline site2/docs/window-functions-context.md:67]: Maybe we can add a "The" at the beginning of this sentence. If you think this is better, then you can change all in this file.

File: site2/docs/window-functions-context.md [hunk 6]
```diff
@@ -0,0 +1,593 @@
...
+
+        return null;
+    }
+}
+```
+<!--END_DOCUSAURUS_CODE_TABS-->
+
+## User config
+
+When you run or update Pulsar Functions created using SDK, you can pass arbitrary key/values to them with the `--user-config` flag. Key/values **must** be specified as JSON.
```
Comment (sijia-w) [inline site2/docs/window-functions-context.md:358]:
```suggestion
When you run or update Pulsar Functions that are created using SDK, you can pass arbitrary key/values to them with the `--user-config` flag. Key/values **must** be specified as JSON. 
```

File: site2/docs/window-functions-context.md [hunk 7]
```diff
@@ -0,0 +1,593 @@
...
+```bash
+bin/pulsar-admin functions create \
+   --user-config '{"word-of-the-day":"verdure"}' \
+  # Other function configs
+ ```
+
+This example accesses values in a Java window function.
+
+`UserConfigFunction` function logs the string `"The word of the day is verdure"` every time the function is invoked (which means every time a message arrives). The user config of `word-of-the-day` is changed **only** when the function is updated with a new config value via 
+multiple ways, such as the command line tool or REST API.
```
Comment (sijia-w) [inline site2/docs/window-functions-context.md:436]: Seems one sentence in two lines here.

File: site2/docs/window-functions-context.md [hunk 8]
```diff
@@ -0,0 +1,593 @@
...
+// Get the whole config map
+Map<String, String> allConfigs = context.getUserConfigMap();
+
+// Get value or resort to default
+String wotd = context.getUserConfigValueOrDefault("word-of-the-day", "perspicacious");
+```
+
+## Routing
+
+When you want to publish as many results as you want, you can use the `context.publish()` interface.
```
Comment (sijia-w) [inline site2/docs/window-functions-context.md:475]:
```suggestion
You can use the `context.publish()` interface to publish as many results as you want.
```

File: site2/docs/window-functions-context.md [hunk 9]
```diff
@@ -0,0 +1,593 @@
...
+```
+<!--END_DOCUSAURUS_CODE_TABS-->
+
+## State storage
+
+Pulsar window functions use [Apache BookKeeper](https://bookkeeper.apache.org) as a state storage interface. StreamNative Platform installation (including the standalone installation) includes the deployment of BookKeeper bookies.
+
+StreamNative Platform integrates with Apache BookKeeper `table service` to store the `state` for functions. For example, the `WordCount` function can store its `counters` state into BookKeeper table service via Pulsar Functions state APIs.
+
+States are key-value pairs, where the key is a string and the value is arbitrary binary data—counters are stored as 64-bit big-endian binary values. Keys are scoped to an individual Pulsar Function and shared between instances of that function.
```
Comment (sijia-w) [inline site2/docs/window-functions-context.md:503]:
```suggestion
States are key-value pairs, where the key is a string and the value is arbitrary binary data—counters are stored as 64-bit big-endian binary values. Keys are scoped to an individual Pulsar function and shared between instances of that function.
```

File: site2/docs/window-functions-context.md [hunk 10]
```diff
@@ -0,0 +1,593 @@
...
+|`getCounter`|Gets the counter value for the key.
+|`putState`|Updates the state value for the key.
+
+You can use the following APIs to access, update, and manage states in Java window functions. 
+
+#### incrCounter
+
+`incrCounter` API increases a built-in distributed counter referred by key.
+
+Applications uses the `incrCounter` API to change the counter of a given `key` by the given `amount`. If the `key` does not exist, a new key is created.
```
Comment (sijia-w) [inline site2/docs/window-functions-context.md:519]:
```suggestion
Applications use the `incrCounter` API to change the counter of a given `key` by the given `amount`. If the `key` does not exist, a new key is created.
```

**Comment by sijia-w** [comment]
@wolfstudy Left my comments, PTAL. Please keep the number and the first letter in "Pulsar Functions", "a function", etc. consistent.

**Comment by sijia-w** [comment]
@Jennifer88huang  Left my comments. Your feedback is appreciated. Thanks.

File: site2/docs/window-functions-context.md [hunk 11]
```diff
@@ -0,0 +1,593 @@
...
+  
+  * Logger object used by the window function, which can be used to create window function log messages.
+
+- [User config](#user-config)
+  
+  * Access to arbitrary user configuration values.
+
+- [Routing](#routing)
+  
+  * Function to publish new messages to arbitrary topics.
```
Comment (Jennifer88huang) [inline site2/docs/window-functions-context.md:28]: What does this sentence mean?

File: site2/docs/window-functions-context.md (hunk 5 above)
Comment (wolfstudy) [reply site2/docs/window-functions-context.md:67]: OK will add it.

File: site2/docs/window-functions-context.md (hunk 9 above)
Comment (wolfstudy) [reply site2/docs/window-functions-context.md:503]: Maybe we should use `Pulsar Function` in here

File: site2/docs/window-functions-context.md [hunk 12]
```diff
@@ -0,0 +1,593 @@
...
+     * Get any user-defined key/value or a default value if none is present.
+     *
+     * @param key
+     * @param defaultValue
+     * @return Either the user config value associated with a given key or a supplied default value
+     */
+    Object getUserConfigValueOrDefault(String key, Object defaultValue);
+```
+
+### Example
```
Comment (Jennifer88huang) [inline site2/docs/window-functions-context.md:366]: which example?

File: site2/docs/window-functions-context.md [hunk 13]
```diff
@@ -0,0 +1,593 @@
...
+        return null;
+    }
+
+}
+```
+<!--END_DOCUSAURUS_CODE_TABS-->
+
+## State storage
+
+Pulsar window functions use [Apache BookKeeper](https://bookkeeper.apache.org) as a state storage interface. StreamNative Platform installation (including the standalone installation) includes the deployment of BookKeeper bookies.
```
Comment (Jennifer88huang) [inline site2/docs/window-functions-context.md:442]: please do not include "streamnative platform" related info in this PR. 
Same comments for the following cases.

File: site2/docs/window-functions-context.md [hunk 14]
```diff
@@ -0,0 +1,532 @@
...
+     * @param amount The amount to be incremented
+     */
+    void incrCounter(String key, long amount);
+```
+
+#### getCounter
+
+The `getCounter` API gets the counter value for the key.
+
+Applications can use the `getCounter` API to retrieve the counter of a given `key` changed by the `incrCounter` API.
```
Comment (Jennifer88huang) [inline site2/docs/window-functions-context.md:477]:
```suggestion
Applications uses the `getCounter` API to retrieve the counter of a given `key` changed by the `incrCounter` API.
```

File: site2/docs/window-functions-context.md [hunk 15]
```diff
@@ -0,0 +1,532 @@
...
+    /**
+     * Update the state value for the key.
+     *
+     * @param key name of the key
+     * @param value state value of the key
+     */
+    void putState(String key, ByteBuffer value);
+```
+
+### Example
```
Comment (Jennifer88huang) [inline site2/docs/window-functions-context.md:505]: Where does this example belongs to?

File: site2/docs/window-functions-context.md (hunk 12 above)
Comment (wolfstudy) [reply site2/docs/window-functions-context.md:366]: Maybe we can delete `Example`?

File: site2/docs/window-functions-context.md (hunk 11 above)
Comment (Jennifer88huang) [reply site2/docs/window-functions-context.md:28]: How about refine it as follows?
Routing is supported in Pulsar Functions. Pulsar Functions send messages to arbitrary topics as per the publish interface.

**Comment by Jennifer88huang** [comment]
@wolfstudy thank you very much for contributing to the content. Could you please add Fixes #5520 in the description? And refine the PR title accordingly?
Concerning line 28, I've left comments, please check, any issue feel free to contact me.

**Comment by Jennifer88huang** [comment]
Great job @wolfstudy. I checked the #5520, the feature was added since v2.3.0. So we need to apply the updates since v2.3.0.

File: site2/website/versioned_docs/version-2.3.0/window-functions-context.md [hunk 16]
```diff
@@ -0,0 +1,528 @@
+---
+id: window-functions-context
+title: Window Functions Context
+sidebar_label: "Window Functions: Context"
+---
```
Comment (Jennifer88huang) [inline site2/website/versioned_docs/version-2.3.0/window-functions-context.md:5]:
```suggestion
---
id: version-2.3.0-window-functions-context
title: Window Functions Context
sidebar_label: "Window Functions: Context"
original_id: window-functions-context
---
```

Comment (Jennifer88huang) [reply site2/website/versioned_docs/version-2.3.0/window-functions-context.md:5]: please check all similar cases.

**Pull request merged**